kernel/src/schema/mod.rs

```diff
@@ -371,47 +527,133 @@ pub struct StructType {
+            fields
+                .iter()
+                .filter(|f| !matches!(f.data_type, DataType::Primitive(_))),
+        )?;
+
+        Ok(Self {
             type_name: "struct".into(),
             fields: fields.into_iter().map(|f| (f.name.clone(), f)).collect(),
-        }
+        })
     }
 
-    pub fn try_new<E>(fields: impl IntoIterator<Item = Result<StructField, E>>) -> Result<Self, E> {
-        let fields: Vec<_> = fields.into_iter().try_collect()?;
-        Ok(Self::new(fields))
+    /// Creates a new [`StructType`] from the given fields without checking for invalid usage of metadata columns.
+    ///
+    /// This should only be used when you are sure that the fields are valid.
+    pub fn new_unchecked(fields: impl IntoIterator<Item = StructField>) -> Self {
+        Self {
+            type_name: "struct".into(),
+            fields: fields.into_iter().map(|f| (f.name.clone(), f)).collect(),
+        }
     }
 
-    /// Get a [`StructType`] containing [`StructField`]s of the given names. The order of fields in
+    /// Gets a [`StructType`] containing [`StructField`]s of the given names. The order of fields in
     /// the returned schema will match the order passed to this function, which can be different
     /// from this order in this schema. Returns an Err if a specified field doesn't exist.
     pub fn project_as_struct(&self, names: &[impl AsRef<str>]) -> DeltaResult<StructType> {
-        let fields = names.iter().map(|name| {
-            self.fields
-                .get(name.as_ref())
-                .cloned()
-                .ok_or_else(|| Error::missing_column(name.as_ref()))
-        });
+        let fields: Vec<_> = names
+            .iter()
+            .map(|name| {
+                self.fields
+                    .get(name.as_ref())
+                    .cloned()
+                    .ok_or_else(|| Error::missing_column(name.as_ref()))
+            })
+            .try_collect()?;
         Self::try_new(fields)
     }
 
-    /// Get a [`SchemaRef`] containing [`StructField`]s of the given names. The order of fields in
+    /// Gets a [`SchemaRef`] containing [`StructField`]s of the given names. The order of fields in
     /// the returned schema will match the order passed to this function, which can be different
     /// from this order in this schema. Returns an Err if a specified field doesn't exist.
     pub fn project(&self, names: &[impl AsRef<str>]) -> DeltaResult<SchemaRef> {
         let struct_type = self.project_as_struct(names)?;
         Ok(Arc::new(struct_type))
     }
 
-    pub fn field(&self, name: impl AsRef<str>) -> Option<&StructField> {
-        self.fields.get(name.as_ref())
+    /// Adds fields to this struct type, returning a new StructType.
+    pub fn add(&self, fields: impl IntoIterator<Item = StructField>) -> DeltaResult<Self> {
+        let mut new_fields: Vec<_> = self.fields.values().cloned().collect();
+        new_fields.extend(fields);
```

Comment:
no need to materialize it:
```suggestion
        let new_fields = self.fields.values().cloned().chain(fields);
```